nativelib/src/main/scala/java/lang/ieee754tostring/ryu/RyuDouble.scala

```diff
@@ -0,0 +1,548 @@
+        (if ((dv == dm &&
+             !(dmIsTrailingZeros && roundingMode.acceptLowerBound(even))) ||
+             (lastRemovedDigit >= 5)) 1
+         else 0)
+    } else {
+      var done = false // workaround break in .java source
+      while ((dp / 10 > dm / 10) && !done) {
+        if ((dp < 100) && scientificNotation) {
+          // Double.toString semantics requires printing at least two digits.
+          done = true
+        } else {
+          lastRemovedDigit = (dv % 10).toInt
+          dp /= 10
+          dv /= 10
+          dm /= 10; removed += 1
+        }
+      }
+      output = dv +
+        (if ((dv == dm || (lastRemovedDigit >= 5))) 1 else 0)
+    }
+    val olength: Int = vplength - removed
+    if (DEBUG) {
+      println("LAST_REMOVED_DIGIT=" + lastRemovedDigit)
+      println("VP=" + dp)
+      println("VR=" + dv)
+      println("VM=" + dm)
+      println("O=" + output)
+      println("OLEN=" + olength)
+      println("EXP=" + exp)
+    }
+
+    // Step 5: Print the decimal representation.
+    // We follow Double.toString semantics here.
+    val result: Array[Char] = Array.ofDim[Char](24)
+    var index: Int          = 0
+    if (sign) {
+      result({ index += 1; index - 1 }) = '-'
+    }
+
+    // Values in the interval [1E-3, 1E7) are special.
+    if (scientificNotation) {
+      for (i <- 0 until olength - 1) {
+        val c: Int = (output % 10).toInt
+        output /= 10
+        result(index + olength - i) = ('0' + c).toChar
+      }
+      result(index) = ('0' + output % 10).toChar
+      result(index + 1) = '.'
+      index += olength + 1
+      if (olength == 1) {
+        result({ index += 1; index - 1 }) = '0'
+      }
+
+      // Print 'E', the exponent sign, and the exponent, which has at most
+      // three digits.
+      result({ index += 1; index - 1 }) = 'E'
+      if (exp < 0) {
+        result({ index += 1; index - 1 }) = '-'
+        exp = -exp
+      }
+      if (exp >= 100) {
+        result({ index += 1; index - 1 }) = ('0' + exp / 100).toChar
+        exp %= 100
+        result({ index += 1; index - 1 }) = ('0' + exp / 10).toChar
+      } else if (exp >= 10) {
+        result({ index += 1; index - 1 }) = ('0' + exp / 10).toChar
+      }
+      result({ index += 1; index - 1 }) = ('0' + exp % 10).toChar
+      new String(result, 0, index)
+    } else {
+      // Otherwise follow the Java spec for values in the interval [1E-3, 1E7).
+      if (exp < 0) {
+        // Decimal dot is before any of the digits.
+        result({ index += 1; index - 1 }) = '0'
+        result({ index += 1; index - 1 }) = '.'
+        var i: Int = -1
+        while (i > exp) {
+          result({ index += 1; index - 1 }) = '0'; i -= 1
+        }
+        val current: Int = index
+        for (i <- 0 until olength) {
+          result(current + olength - i - 1) = ('0' + output % 10).toChar
+          output /= 10; index += 1
+        }
+      } else if (exp + 1 >= olength) {
+        for (i <- 0 until olength) {
```

Comment:
Use while loops instead of for loops. While interflow can optimize closures away, I'd not risk it in performance critical performance-sensitive code like that.

Reply:
Based on discussion on gitter and #1439, this comment is out of date. For loops over ranges are completely free.

Reply:
Just to be clear, free even when the bound (bounds?) of the range is a variable, known only at runtime?  Your last comment implies that but I want to double check.

Reply:
Yep, should be all good. For example for:
```
  @noinline def loop2(a: Int, b: Int): Unit = {
    for (i <- a until b) {
      f(i)
    }
  }
```
We get:
```
__SM5Test$D5loop2iiuE:
00000001000423f0	pushq	%rbp
00000001000423f1	pushq	%rbx
00000001000423f2	pushq	%rax
00000001000423f3	movl	%edx, %ebx
00000001000423f5	movl	%esi, %ebp
00000001000423f7	cmpl	%ebx, %ebp
00000001000423f9	cmovgel	%ebp, %ebx
00000001000423fc	jge	0x10004240d
00000001000423fe	nop
0000000100042400	movl	%ebp, %esi
0000000100042402	callq	__SM5Test$KD1fiuEX5Test$iE
0000000100042407	incl	%ebp
0000000100042409	cmpl	%ebp, %ebx
000000010004240b	jne	0x100042400
000000010004240d	leaq	
```